R/labeller.r

```diff
@@ -566,126 +533,57 @@ build_strip <- function(label_df, labeller, theme, horizontal) {
   }
 }
 
-#' Create list of strip labels
-#'
-#' Calls [title_spec()] on all the labels for a set of strips to create a list
-#' of text grobs, heights, and widths.
-#'
-#' @param labels Matrix of strip labels
-#' @param element Theme element (see [calc_element()]).
-#' @param gp Additional graphical parameters.
-#'
-#' @noRd
-create_strip_labels <- function(labels, element, gp) {
-  grobs <- lapply(labels, title_spec,
-    x = NULL,
-    y = NULL,
-    hjust = element$hjust,
-    vjust = element$vjust,
-    angle = element$angle,
-    gp = gp,
-    debug = element$debug
-  )
-  dim(grobs) <- dim(labels)
-  grobs
-}
-
 #' Grob for strip labels
 #'
 #' Takes the output from title_spec, adds margins, creates gList with strip
 #' background and label, and returns gtable matrix.
 #'
-#' @param grobs Output from [title_spec()].
+#' @param grobs Output from [titleGrob()].
 #' @param theme Theme object.
-#' @param element Theme element (see [calc_element()]).
-#' @param gp Additional graphical parameters.
 #' @param horizontal Whether the strips are horizontal (e.g. x facets) or not.
 #' @param clip should drawing be clipped to the specified cells (‘"on"’),the
 #'   entire table (‘"inherit"’), or not at all (‘"off"’).
 #'
 #' @noRd
-ggstrip <- function(grobs, theme, element, gp, horizontal = TRUE, clip) {
+assemble_strips <- function(grobs, theme, horizontal = TRUE, clip) {
+  if (length(grobs) == 0 || is.zero(grobs[[1]])) return(grobs)
+
+  # Add margins to non-titleGrobs so they behave eqivalently
+  grobs <- lapply(grobs, function(g) {
+    if (inherits(g, "titleGrob")) return(g)
+    add_margins(g, grobHeight(g), grobWidth(g), margin_x = TRUE, margin_y = TRUE)
```

Comment:
This line needs one more modification: The first argument to `add_margins()` has to be `gList(g)` instead of just `g`. It turns out the documentation to `add_margins()` is incorrect (and probably should be fixed). It takes a `gList` as an argument, not a grob.

Here is a reprex that doesn't require ggtext:
``` r
library(ggplot2)
library(rlang)

element_txt <- function(...) {
  out <- element_text(...)
  class(out) <- c("element_txt", "element_text", "element")
  out
}

element_grob.element_txt <- function(element, label = "", x = NULL, y = NULL, ...) {
  if (is.null(label))
    return(zeroGrob())
  
  x <- x %||% element$hjust
  y <- y %||% element$vjust
  
  element_gp <- grid::gpar(
    fontsize = element$size, col = element$colour,
    fontfamily = element$family, fontface = element$face,
    lineheight = element$lineheight
  )
  
  grid::textGrob(label, x, y, hjust = element$hjust, vjust = element$vjust, gp = element_gp)
}

df <- data.frame(x = 1:3, y = 1:3, a = letters[1:3])

ggplot(df, aes(x, y)) + 
  geom_point() +
  facet_wrap(~a) +
  theme(axis.text.x = element_txt(colour = "red"))
```

![](https://i.imgur.com/AVYk3qj.png)

``` r

ggplot(df, aes(x, y)) + 
  geom_point() +
  facet_wrap(~a) +
  theme(strip.text = element_txt(colour = "red"))
#> Error in setChildren(gt, children): 'children' must be a "gList"
```

<sup>Created on 2020-01-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I think it would be good to have a unit test for this code path, maybe based on this simple reprex I wrote.